### DISCUSSION TITLE
OpenRadioss parallel

### DISCUSSION THREAD

**Discussion by CinekPoland** [OpenRadioss/OpenRadioss#1099] [Q&A]
Hi. I run OpenRadioss in using parallel option like this 

cd C:\Program Files (x86)\Intel\oneAPI\env
call vars.bat

cd C:\Program Files\OpenRadioss\exec
set OMP_NUM_THREADS=32
commands.bat
starter_win64.exe -i C:\Users\cxxxxx\Desktop\bird\BIRD_WINDSHIELD_v1_0000.rad -np 1 -outfile=".\res"

mpiexec -n 32  engine_win64.exe -i C:\Users\xxxxx\Desktop\bird\BIRD_WINDSHIELD_v1_0001.rad  -outfile=".\res"

I found some informations which are included below
![1](https://user-images.githubusercontent.com/119892611/217637446-f8ec2f26-40d1-4401-b91f-131d95762dd8.jpg)

What is exactly "-np"  ? What is " P MPI processes" ? I understand what is thread, I know what is "domain" but I don't understand what is " P MPI processes"

**Comment by MarianBulla** [comment]
Hi, 
the more detailled information can be found on the website directly from Intel (and further subsites, here): https://www.intel.com/content/www/us/en/develop/documentation/get-started-with-mpi-for-windows/top.html

Hope it helps you ?

Thanks and best regards,
Marian

**Reply by CinekPoland** [reply]
On this website I don't see any information about "P MPI processes"

**Reply by elequiniou** [reply]
Hi, here in the help it only means if one uses "-np P " it will create "P MPI" processes where P is any integer. At first, I would recommend to set P as the number of physical cores on your machine. And let N=1 (N=number of OpenMP threads). So in your case:

set OMP_NUM_THREADS=1

starter_win64.exe -i C:\Users\cxxxxx\Desktop\bird\BIRD_WINDSHIELD_v1_0000.rad -np 32 -outfile=".\res"

mpiexec -n 32 engine_win64.exe -i C:\Users\xxxxx\Desktop\bird\BIRD_WINDSHIELD_v1_0001.rad -outfile=".\res"

It will create 32 MPI domains by Starter and mpiexec will launch 32 Engine processes.  Hope this clarifies and helps.

**Reply by CinekPoland** [reply]
Thank you for you answer

**Comment by CinekPoland** [comment]
unfortunately doesn't work . 
![1](https://user-images.githubusercontent.com/119892611/220762481-129afa61-f690-4e02-aadd-ce93d4821160.jpg)
![2](https://user-images.githubusercontent.com/119892611/220762483-97728c03-908e-4433-9a99-81fc82f49db7.jpg)
![3](https://user-images.githubusercontent.com/119892611/220762487-cba2f867-eb66-41a9-9cff-6769093819a4.jpg)

**Comment by OlivierWienholtz** [comment]
Hello,

There is an issue on the command line : please use the intel MPI binary when running MPI programs : 
The executable to use is : engine_win64_impi

He executed 32 times the SMP Engine with mpiexec (so 32 jobs of 1 MPI process) and not 1 job of 32 MPI processes.

Hope it helps,
Olivier

**Comment by CinekPoland** [comment]
Hi. Thank you for your answer but how can i use engine_win64_impi ? Start engine_win64_impi using command line (cmd) ?

**Comment by MarianBulla** [comment]
Hi, 
please see the screenshot, how I'm starting OpenRadioss using MPI.
Hope it helps you ?
![image](https://user-images.githubusercontent.com/105412502/222363690-cadca8fa-c2dc-4cdf-8f22-3b9d028fb8b5.png)

**Reply by CinekPoland** [reply]
are three three commands enough? Another question. Can I use different material models like the Johnson Cook model for sph simulation?

**Comment by OlivierWienholtz** [comment]
Hello
Here comes the Commands for Windows: 
You can also have a look in the HowTo Run OpenRadioss document : 
https://github.com/OpenRadioss/OpenRadioss/blob/main/INSTALL.md

1. You need o Intel MPI 2021.8/OneAPI 2023.0.0  Installation 

3. In Your shell :  Call thee environment variables to Load OneAPI Intel MPI 2021.8 : 
    call [Path to Intel OneAPI]\env\vars.bat
   If you have a full OneAPI 2023.0.0 installation, you can also the INTEL ONE API Command prompt from Intel Menu Folder.

4. Set The number of threads for yiur Hybrid Run : 
set OMP_NUM_THREADS=[N]

5. Launch Starter & Engine.
  starter_win64.exe -i [Starter input file] -np [P]
  mpirun -delegate -np [P]  engine_win64.exe  -i [Engine input file]

Hope it helps,
Best Regards,
Olivier W.

**Reply by CinekPoland** [reply]
Thank you for you answer Olivier.  My machine has two physical CPU , each cpu has 16 core, also 64GB  has my machine. MPI+OpenMP is the the best option in my case ? 

I did Intel OneApi update like you see, then a tried commands which you wrote in your last post.  And like you see doesn't work. Maybe the problem is in the installation method? 
![1](https://user-images.githubusercontent.com/119892611/222928319-fc98bfdd-7702-4e75-b2ea-8e2dd76c9690.jpg)

**Comment by OlivierWienholtz** [comment]
Hello Cinek,
With 2 CPUS & 16 cores, I would recommend full MPI runs.


I can see that you are in the OneAPI installation directory.
It won't be possible to write any result files there, the System did not find the Starter.

In this shell where you loaded the IntelMPI variables:
I would recommend you to enter in a directory where you can write 

cd \Users\cinek\Desktop\bird\

Then set environment the Variables for Windows:
Pls See here : https://github.com/OpenRadioss/OpenRadioss/blob/main/INSTALL.md


    set OPENRADIOSS_PATH=[Path to OpenRadioss root directory / Windows Style]
    set RAD_CFG_PATH=%OPENRADIOSS_PATH%\hm_cfg_files
    set RAD_H3D_PATH=%OPENRADIOSS_PATH%\extlib\h3d\lib\win64
    set KMP_STACKSIZE=400m
    set PATH=%OPENRADIOSS_PATH%\extlib\hm_reader\win64;%PATH% 

To run Starter & Engine under Windows :
16  MPI processes &  1 Threads

set OMP_NUM_THREADS=1]
  starter_win64.exe -i [Starter input file] -np 16
  mpirun -delegate -np 16  engine_win64.exe  -i [Engine input file]

Let Me know if this works,

Best Regards,
Olivier

**Reply by CinekPoland** [reply]
![4](https://user-images.githubusercontent.com/119892611/223852736-81b3db1c-90c8-4324-a367-e6b80ad93425.jpg)
![5](https://user-images.githubusercontent.com/119892611/223852744-b4c61f17-8405-416c-9a55-01b31ab58add.jpg)

The first few commands works I think (no errors) but starter doesn't.

**Comment by CinekPoland** [comment]
![2](https://user-images.githubusercontent.com/119892611/223232653-75754426-952f-427f-bda6-32f2352921d4.jpg)
![3](https://user-images.githubusercontent.com/119892611/223232656-afecfc9a-b11c-4fe4-8600-feef298b9d14.jpg)
The first few commands works I think (no errors) but starter doesn't.

**Reply by MarianBulla** [reply]
I can see, that you are using special characters (here it is the blanck or space) within your path variable. 
You have to put the entire path in " " (quotation marks) then.
Hope it helps you ?
Thanks and best 
Marian
![image](https://user-images.githubusercontent.com/105412502/223381762-8ca1ecec-b708-47c3-bad1-d0ad1f69fecb.png)

**Comment by MarianBulla** [comment]
Hi, 
please make sure, you have set all the necessary path variables:
![image](https://user-images.githubusercontent.com/105412502/223380695-4fd052f1-9da8-4169-9522-19a51a039d65.png)
Hope it helps you ?
Thanks and best regards,
Marian

**Reply by CinekPoland** [reply]
![2](https://user-images.githubusercontent.com/119892611/223852835-4c72cd3e-d050-44a5-943a-883df9457b2d.jpg)
![3](https://user-images.githubusercontent.com/119892611/223853333-84beac60-c4a1-444e-82d6-63274adcba0e.jpg)

Should I download "mkl_intel_thread.2.dll" ? 
or should I uninstall intel one api 2022 ? I have 2022 and 2023 version 
While I was installing openradioss I didn't use  Cyngwin.

**Reply by MarianBulla** [reply]
Hi,
there is no need for using Cygwin.
It looks like you have still issues with the PATH environment variable.

Can you double-check, what is in your PATH variable by typing "echo %PATH%", please ?
it should point to the executables of OpenRadioss AND to the external libraries:
![image](https://user-images.githubusercontent.com/105412502/223951077-66bae666-2ead-4064-8dc6-353ef83c1f42.png)

**Reply by CinekPoland** [reply]
![2](https://user-images.githubusercontent.com/119892611/224128923-4ef7d676-e86b-4e74-9cd4-fa0d9ea1504b.jpg)

**Reply by MarianBulla** [reply]
Can you make sure, you have add the path to your OpenRadioss executables, please (see screenshot of my windows executables) ?
![image](https://user-images.githubusercontent.com/105412502/224252496-beac7c0d-8ac1-45f1-b888-d25e41f92030.png)
![image](https://user-images.githubusercontent.com/105412502/224252625-fb71390c-b834-416a-8906-137db0e03642.png)

**Reply by MarianBulla** [reply]
This is how I add the enviroment variables, like PATH:
![image](https://user-images.githubusercontent.com/105412502/224252899-2b23596c-ef8f-4a93-989c-5e0f3e6dac3e.png)

**Reply by CinekPoland** [reply]
![2](https://user-images.githubusercontent.com/119892611/224413017-4c8f3c86-cce1-47cc-ac26-758b95a87bac.jpg)
![3](https://user-images.githubusercontent.com/119892611/224413023-1a94be89-48b1-48fd-b16f-512e0ed31d5c.jpg)
![5](https://user-images.githubusercontent.com/119892611/224413025-5faec604-021e-437e-a788-52a4dd67b996.jpg)
![6](https://user-images.githubusercontent.com/119892611/224413026-e90ab4aa-2808-4f87-a202-7fe6a8b67ef5.jpg)
The last two photos I added are almost the same like last photo of your last post

**Reply by CinekPoland** [reply]
before command "starter i..." I  wrote all commands almost like commands in your last post. What the command "starter i..."  should looks like ?

**Reply by MarianBulla** [reply]
There is an issue with your OPENRADIOSS_PATH variable.
Can you move your entire OpenRadioss folder into a different folder of your c-drive, without any special character in the name (the blank is a special character), e.g. "C:\test\OpenRadioss" and readjust the OPENRADIOSS_PATH, please ?
![image](https://user-images.githubusercontent.com/105412502/224420169-e00bfaf4-254f-4a17-834d-fa9fc49b7a3f.png)

**Reply by MarianBulla** [reply]
For the starter command, it should look like this:
![image](https://user-images.githubusercontent.com/105412502/224422116-168c51bd-e7fa-4922-97ff-90ad0bc9b42f.png)

**Reply by CinekPoland** [reply]
I copied folder openradioss to C:\test but doesn't works using old command so I think I have to reinstall openradioss. 
Files like 0000.rad and 0001.rad  must be in the same folder like starter.win64.exe ? 
I guess you run openradioss fully MPI mode  ?

**Reply by MarianBulla** [reply]
Please see, how I startet OpenRadioss from a fresh opened command window:
![image](https://user-images.githubusercontent.com/105412502/224727210-7dcf07c7-e19e-4cf1-9f67-bfdfde7a3ff4.png)

**Comment by OlivierWienholtz** [comment]
Hi Cinek, 

You don't need to copy binaries in executation folder : eg where the input file is : 
you have direct access when you add the directory to OpenRadioss Installation in your PATH variable 

See Marian's the screenshot up: setting OPENRADIOSS_PATH and PATH does the trick : 

assuming you have installed OpenRadioss in c:\Program Files
set OPENRADIOSS_PATH="c:\Program Files\OpenRadioss" 
set PATH=%OPENRADIOSS_PATH%\exec;%PATH% 

Hope it helps,
Best Regards
Olivier W.
